props/saber_BC_buttons.h

```diff
@@ -386,31 +1161,254 @@ EFFECT(volmax);     // for maximum volume reached
+      current_volume_ += max_volume_ * 0.10;
+      if (current_volume_ > max_volume_) current_volume_ = max_volume_;
+      dynamic_mixer.set_volume(current_volume_);
+      mode::getSL<SPEC>()->SayVolumeUp();
+      announced_ = false;
+    } else if (!announced_) {
+      QuickMaxVolume();
+      announced_ = true;
+    }
+  }
+
+  void prev() override {
+    int current_volume_ = dynamic_mixer.get_volume();
+    if (current_volume_ > min_volume_) {
+      current_volume_ -= max_volume_ * 0.10;
+      if (current_volume_ < min_volume_) current_volume_ = min_volume_;
+      dynamic_mixer.set_volume(current_volume_);
+      mode::getSL<SPEC>()->SayVolumeDown();
+      announced_ = false;
+    } else if (!announced_) {
+      QuickMinVolume();
+      announced_ = true;
+    }
+  }
+
+  void QuickMaxVolume() {
+    dynamic_mixer.set_volume(max_volume_);
+    PVLOG_NORMAL << "** Maximum Volume\n";
+    mode::getSL<SPEC>()->SayMaximumVolume();
+  }
+
+  void QuickMinVolume() {
+    dynamic_mixer.set_volume(min_volume_);
+    PVLOG_NORMAL << "** Minimum Volume\n";
+    mode::getSL<SPEC>()->SayMinimumVolume();
+  }
+
+  void update() override {  // Overridden to substitute the tick sound
+    float volume = dynamic_mixer.get_volume();
+    percentage_ = round((volume / max_volume_) * 10) * 10;
+    SaberBase::DoEffect(EFFECT_VOLUME_LEVEL, 0);
+  }
+
+  void select() override {
+    PVLOG_NORMAL << "** Saved - Exit Volume Menu\n";
+    mode::getSL<SPEC>()->SaySave();
+    SPEC::SteppedMode::select();
+  }
+
+  void exit() override {
+    PVLOG_NORMAL << "** Cancelled - Exit Volume Menu\n";
+    percentage_ = initial_percentage_;
+    dynamic_mixer.set_volume(initial_volume_);
+    mode::getSL<SPEC>()->SayCancel();
+    SPEC::SteppedMode::exit();
+  }
+
+  bool mode_Event2(enum BUTTON button, EVENT event, uint32_t modifiers) override {
+    switch (EVENTID(button, event, 0)) {
+      // Custom button controls for BCVolumeMode
+      case EVENTID(BUTTON_POWER, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMaxVolume();
+        return true;
+
+      case EVENTID(BUTTON_POWER, EVENT_SECOND_HELD_MEDIUM, 0):
+      case EVENTID(BUTTON_AUX, EVENT_FIRST_HELD_MEDIUM, 0):
+        QuickMinVolume();
+        return true;
+    }
+    // Use the select and exit controls from SelectCancelMode
+    return SPEC::SelectCancelMode::mode_Event2(button, event, modifiers);
+  }
+};
+
+#ifdef DYNAMIC_BLADE_LENGTH
+
+template<class SPEC>
+struct BCSelectBladeMode : public SPEC::MenuBase {
+  int current_blade() { return this->pos_ + 1; }
+  ShowColorSingleBladeTemplate<Pulsing<White,Black,800>,Pulsing<White,Black,800>> highlighted_blade_;
+  bool single_blade_adjusted_ = false;
+  uint16_t size() override { return NUM_BLADES; }
+
+  void mode_activate(bool onreturn) override {
+    // If only one blade, jump right to ChangeBladeLengthMode
+    // If just adjusted and returning here via popMode, just exit.
+    if (size() == 1) {
+      if (!single_blade_adjusted_) {
```

Comment:
I think there are at least three better ways to do this....

1. You can use the "onreturn" variable to detect that we got here by popping rather than by pushing
2. you could pop this mode before pushing the ChangeBladeLengthMode so it's not on the mode stack at all anymore.
3. you could use #if NUM_BLADES == 1 and where this mode is used and use ChangeBladeLengthMode directly if that is the case.

Reply:
oh wow `onreturn` has been there the whole time huh.
I was doing things in Volume menu like "first_time_through_" or whatever and this is readily available to do that exact job. So nice, thanks.
Adjusted.